beacon-chain/sync/pending_attestations_queue.go

```diff
@@ -91,52 +91,61 @@ func (s *Service) processPendingAtts(ctx context.Context) error {
 	return s.sendBatchRootRequest(ctx, pendingRoots, randGen)
 }
 
-func (s *Service) processAttestations(ctx context.Context, attestations []ethpb.SignedAggregateAttAndProof) {
+func (s *Service) processAttestations(ctx context.Context, attestations []any) {
 	for _, signedAtt := range attestations {
-		att := signedAtt.AggregateAttestationAndProof().AggregateVal()
-		// The pending attestations can arrive in both aggregated and unaggregated forms,
-		// each from has distinct validation steps.
-		if att.IsAggregated() {
-			s.processAggregated(ctx, signedAtt)
-		} else {
-			s.processUnaggregated(ctx, att)
+		// The pending attestations can arrive as both aggregates and attestations,
+		// and each form has to be processed differently.
+		switch t := signedAtt.(type) {
+		case ethpb.Att:
+			s.processAtt(ctx, t)
+		case ethpb.SignedAggregateAttAndProof:
+			s.processAggregate(ctx, t)
+		default:
+			log.Debugf("Unexpected attestation type %T in pending attestation queue. Attestation will not be processed", t)
 		}
 	}
 }
 
-func (s *Service) processAggregated(ctx context.Context, att ethpb.SignedAggregateAttAndProof) {
-	aggregate := att.AggregateAttestationAndProof().AggregateVal()
+func (s *Service) processAggregate(ctx context.Context, aggregate ethpb.SignedAggregateAttAndProof) {
+	att := aggregate.AggregateAttestationAndProof().AggregateVal()
 
 	// Save the pending aggregated attestation to the pool if it passes the aggregated
 	// validation steps.
-	valRes, err := s.validateAggregatedAtt(ctx, att)
+	valRes, err := s.validateAggregatedAtt(ctx, aggregate)
 	if err != nil {
 		log.WithError(err).Debug("Pending aggregated attestation failed validation")
 	}
 	aggValid := pubsub.ValidationAccept == valRes
-	if s.validateBlockInAttestation(ctx, att) && aggValid {
+	if s.validateBlockInAttestation(ctx, aggregate) && aggValid {
 		if features.Get().EnableExperimentalAttestationPool {
-			if err = s.cfg.attestationCache.Add(aggregate); err != nil {
-				log.WithError(err).Debug("Could not save aggregate attestation")
+			if err = s.cfg.attestationCache.Add(att); err != nil {
+				log.WithError(err).Debug("Could not save aggregated attestation")
 				return
 			}
 		} else {
-			if err := s.cfg.attPool.SaveAggregatedAttestation(aggregate); err != nil {
-				log.WithError(err).Debug("Could not save aggregate attestation")
-				return
+			if att.IsAggregated() {
```

Comment:
A `SignedAggregateAttAndProof` can hold an attestation with just one bit set, in which case saving the attestation through `SaveAggregatedAttestation` will fail